src/Lib.hs

```diff
@@ -1,56 +1,97 @@
+{-# LANGUAGE OverloadedStrings   #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeOperators       #-}
 
 module Lib
   ( User(..)
   , myServer
   , myApi
   ) where
 
-import           Data.Aeson               as JSON
-import           Data.Foldable            (find)
-import           Models                   (User (..))
+import           Control.Monad.IO.Class     (MonadIO (liftIO))
+import           Data.Functor               (void)
+import           Data.Pool                  (Pool, withResource)
+import           Database.PostgreSQL.Simple (Connection, Only (Only), execute,
+                                             query, query_)
+import           Models                     (User (..))
 import           Servant
 
 type UsersAPI
   = "users" :> "list-all" :> Get '[ JSON] [User] 
       :<|> "users" :> Capture "usrId" Int :> Get '[ JSON] User 
-      :<|> "createUser" :> ReqBody '[ JSON] User :> Post '[ JSON] User 
-      :<|> "deleteUser" :> Capture "usrId" Int :> Delete '[ JSON] [User] 
-      :<|> "updateUser" :> Capture "usrId" Int :> ReqBody '[ JSON] User :> Put '[ JSON] [User]
+      :<|> "deleteUser" :> Capture "usrId" Int :> Delete '[ JSON] NoContent 
+      :<|> "upsetUser" :> ReqBody '[ JSON] User :> Put '[ JSON] NoContent
 
-myUsers :: [User]
-myUsers = [User 0 "charly" $ Just 25, User 1 "gon" $ Just 22]
+getUserById :: Pool Connection -> Int -> IO (Maybe User)
+getUserById conns uid =
+  handleDbResponse
+    <$> withResource
+          conns
+          (\conn ->
+             query
+               conn
+               "select * from users where id = ?"
+               (Only uid :: Only Int))
+  where
+    handleDbResponse userResponse =
+      case userResponse of
+        []    -> Nothing
+        (x:_) -> Just x
 
--- TODO: move to utils
-getUserById :: Int -> Maybe User
-getUserById usrId = find (\user -> userId user == usrId) myUsers
+deleteUserById :: Pool Connection -> Int -> IO NoContent
+deleteUserById conns uid = do
+  _ <-
+    withResource
+      conns
+      (\conn ->
+         execute conn "delete from users where id = ?" (Only uid :: Only Int))
+  return NoContent
 
--- TODO: move to utils
-deleteUserFromList :: Int -> [User]
-deleteUserFromList uid = filter (\user -> userId user /= uid) myUsers
+createUserInDb :: Pool Connection -> User -> IO NoContent
+createUserInDb conns newUser = do
+  void
+    $ withResource
+        conns
+        (\conn ->
+           execute
+             conn
+             "insert into users (id, name, age) values (?, ?, ?)"
+             newUser)
+  return NoContent
 
--- Not super happy with this approach but I will change it when setting DB connection
-updateUserFromList :: Int -> User -> [User]
-updateUserFromList uid updatedUser = updatedUser : deleteUserFromList uid
+updateUserInDb :: Pool Connection -> User -> IO NoContent
+updateUserInDb conns userData = do
+  void $ deleteUserById conns (userId userData)
+  void $ createUserInDb conns userData
+  return NoContent
```

Comment:
For creation and updating operations, it's usually a good idea to return the entity that is created/updated. The purpose of this is twofold: it communicates to the caller of the function about the changes in state that were actually performed to the requested entity, and it also allows us to perform subsequent operations on the updated resource without having to retrieve it explicitly after the creation/update operation.

The same philosophy applies to endpoints: it's usually good to return the resource that you are creating or updating in POST/PUT endpoint.